### ISSUE TITLE
Trying to 'docker service create' w/ non-existent image never times out, chews CPU

### ISSUE THREAD

**Issue by nathanleclaire** [docker/docker#24082]
**Output of `docker version`:**

```
$ docker version
Client:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   906eacd
 Built:        Fri Jun 17 20:35:33 2016
 OS/Arch:      darwin/amd64
 Experimental: true

Server:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   a7119de
 Built:        Fri Jun 17 22:09:20 2016
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
$ docker info
Containers: 64
 Running: 1
 Paused: 0
 Stopped: 63
Images: 271
Server Version: 1.12.0-rc2
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 522
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host overlay null
Swarm: active
 NodeID: 1iw918nblauer800gaxwtcyty
 IsManager: Yes
 Managers: 1
 Nodes: 1
 CACertHash: sha256:859d2b3ba1360713500512897ff6b6df1237975d7b79bfc57319daefc3efe72b
Runtimes: default
Default Runtime: default
Security Options: seccomp
Kernel Version: 4.4.13-moby
Operating System: Alpine Linux v3.4
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 1.954 GiB
Name: moby
ID: C2PY:RMJT:LQIX:BQFA:DENP:OIQC:6VCU:DID4:27BT:PAAM:IUEF:EULO
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 48
 Goroutines: 121
 System Time: 2016-06-28T23:41:44.584015126Z
 EventsListeners: 1
No Proxy: *.local, 169.254/16
Username: nathanleclaire
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Docker for Mac, latest beta

**Steps to reproduce the issue:**

Note there are no services and resource usage of the Docker host is minimal.

```console
$ docker service ls
ID  NAME  REPLICAS  IMAGE  COMMAND
```

![screen shot 2016-06-28 at 4 41 17 pm](https://cloud.githubusercontent.com/assets/1476820/16436503/37c3b4e6-3d53-11e6-9906-f006e24e06a6.png)

Now try to run a service with a decent amount of replicas with an image that does not exist.  This might happen if for instance you mis-type something or your automation fails / is buggy.  Or if the network is unreliable.

```console
$ docker service create --replicas 20 ngnix
2a8b7lbh1ncdh2rksyfe3ga09
```

**Describe the results you received:**

Swarm mode spins forever trying to schedule it, check out this CPU usage (this is with only that faulty service running).

![screen shot 2016-06-28 at 4 37 10 pm](https://cloud.githubusercontent.com/assets/1476820/16436565/a8b49e2c-3d53-11e6-8c14-668820d6c4a3.png)

**Describe the results you expected:**

`docker` should time out at some point trying to pull this faulty image?  At the very least, it should not chew up so much CPU.

Consider the analogy of a client trying to access a network endpoint which does not exist, for instance.  Do most network clients sit forever trying to access an endpoint when it is not available?  Or do they time out after a certain amount of time failing to make the request?

Additionally, what happens if you attempt to `docker service create` with no network available?  Better to fail fast than seem to succeed and put the burden of understanding why on the user IMO.

**Cute animal picture**

![](http://weneedfun.com/wp-content/uploads/2015/05/cute-panda-5.jpg)

**Comment by aaronlehmann** [comment]
See https://github.com/docker/swarmkit/issues/1015

**Comment by stevvooe** [comment]
> "Pray, Mr. Babbage, if you put into the machine wrong figures, will the right answers come out?"

No, and I'll use all your CPUs. 🐹 

But, let me give you a new rhyme to live by:

> SwarmKit doesn't quit.

> Consider the analogy of a client trying to access a network endpoint which does not exist, for instance. Do most network clients sit forever trying to access an endpoint when it is not available? Or do they time out after a certain amount of time failing to make the request?

In this analogy, if my upstream goes down, I have to get out of bed (figuratively, maybe systemd does it for me) and restart the process until the network is available again. With the approach in swarm mode, when the upstream recovers, the service will converge and everything will work again. This means I can deploy my service before my image or even before I deploy my registry.

docker/swarmkit#1015 is apt, but I do agree this is using too much CPU. I suspect we need a little back off when re-trying tasks. We deferred this to the orchestrator in https://github.com/docker/swarmkit/pull/978 and I suspect we just need to backoff there as failures pile up, rather blast the nodes with new tasks.

A circuit breaker based failure detection algorithm with periodic tests would work well here:

1. Encounter bunch of failures.
2. Hang out for awhile.
3. Dispatch a small number of tasks to see if they are successful.
4. If successful, dispatch everything, if not, return to 2.

cc @aaronlehmann 

In the mean time, could you grab a CPU profile?

**Comment by nathanleclaire** [comment]
> In the mean time, could you grab a CPU profile?

Happy to -- Does swarmkit / Docker integration have some ways to do this out of the box? (I think I recall someone mentioning swarmkit might make pprof information accessible through an endpoint?)

**Comment by aaronlehmann** [comment]
@nathanleclaire: There is a parameter called "restart delay" (`--restart-delay` in `service create`) which controls how long the orchestrator will wait before restarting a failed task. The default is to pause 5 seconds between attempts, and I just confirmed this behavior in -rc2. I'm not seeing it chew much CPU because of the 5 second delay.

Currently we don't do backoff beyond this fixed delay, but it's something to consider.

**Comment by nathanleclaire** [comment]
Here's pprof output: https://www.dropbox.com/s/369zqxyoc201zqc/pprof.localhost%3A8080.samples.cpu.005.pb.gz?dl=1

```console
$ go tool pprof ~/Downloads/pprof.localhost-8080.samples.cpu.005.pb.gz
Entering interactive mode (type "help" for commands)
(pprof) top5 -cum
0.06s of 9.40s total ( 0.64%)
Dropped 460 nodes (cum <= 0.05s)
Showing top 5 nodes out of 274 (cum >= 1.81s)
      flat  flat%   sum%        cum   cum%
         0     0%     0%      7.33s 77.98%  runtime.goexit
         0     0%     0%      3.25s 34.57%  github.com/docker/swarmkit/agent.(*Agent).run
     0.02s  0.21%  0.21%      3.25s 34.57%  github.com/docker/swarmkit/agent.(*worker).Assign
     0.04s  0.43%  0.64%      1.90s 20.21%  github.com/docker/swarmkit/agent.(*taskManager).run
         0     0%  0.64%      1.81s 19.26%  runtime.mcall
(pprof)
```

**Comment by nathanleclaire** [comment]
graph as well if you are interested :)

![pprof](https://cloud.githubusercontent.com/assets/1476820/16532439/d1507ad8-3f86-11e6-8392-b080db8b1297.png)

**Comment by nathanleclaire** [comment]
The CPU usage happens in spikes, I'm guessing once every 5 seconds due to the setting @aaronlehmann mentioned

**Comment by aluzzardi** [comment]
@nathanleclaire CPU of the worker or the manager?

**Comment by nathanleclaire** [comment]
@aluzzardi In this case, it's both -- just one node

**Comment by aluzzardi** [comment]
@nathanleclaire Any way to reproduce this on a 2 machine swarm with the manager being drained?

The goal would be to determine if the spike is happening at the orchestration layer (manager) or execution layer (worker).

Also, by any chance, is this happening in AWS? In that case a screenshot of the CPU graph would really help determine if the spikes map the restart attempts

**Comment by nathanleclaire** [comment]
The displayed profile is on Docker for Mac, but sure I'd be happy to try it out on AWS as well.

**Comment by aluzzardi** [comment]
@nathanleclaire awesome, thanks! Out of curiosity: how did you fetch a pprof profile out of Docker for Mac? Is the engine started in debug mode?

**Comment by stevvooe** [comment]
> how did you fetch a pprof profile out of Docker for Mac?

We should really have a button for that. @justincormack @dsheets @FrenchBen

**Comment by nathanleclaire** [comment]
>  @nathanleclaire awesome, thanks! Out of curiosity: how did you fetch a pprof profile out of Docker for Mac? Is the engine started in debug mode?

Somewhat curiously, the daemon doesn't seem to be started with debug enabled.  To get the profile I followed instructions [here](https://gist.github.com/kolyshkin/bd25e9de97954b330d8a), sorta.

What I did:

In one window, forward D4M unix socket proxy to `localhost:8080` (not sure if pprof can connect directly to unix socket, it might be do-able):

```console
$ socat -d -d TCP-LISTEN:8080,fork,bind=localhost UNIX:/var/run/docker.sock
```

Then `go tool pprof` on `localhost:8080/debug/pprof/profile`.

```console
$ go tool pprof http://localhost:8080/debug/pprof/profile
Fetching profile from http://localhost:8080/debug/pprof/profile
Please wait... (30s)
Saved profile in /Users/nathanleclaire/pprof/pprof.localhost:8080.samples.cpu.007.pb.gz
Entering interactive mode (type "help" for commands)
(pprof) exit
```

As you can see it drops the `pprof` file in a subdirectory of `~` which can be imported by `go tool pprof` later for further analysis.

e.g., you could export a graph like above if you have `dot` / graphviz installed:

```console
$ go tool pprof -png ~/pprof/pprof.localhost\:8080.samples.cpu.007.pb.gz >007.png
```

It probably wouldn't be too bad to throw together a little container to get all of this to happen automatically.

**Comment by aluzzardi** [comment]
@nathanleclaire Thanks!

**Comment by nathanleclaire** [comment]
👍

**Comment by stevvooe** [comment]
You can also collect them with curl:

```console
$ curl --unix-socket /var/run/docker.sock http:/debug/pprof/goroutine\?debug\=2
```

**Comment by nathanleclaire** [comment]
Oh nice, hadn't realized getting the raw dump was that simple.

**Comment by thaJeztah** [comment]
we should document that somewhere :-)

**Comment by nathanleclaire** [comment]
FWIW having this loop running on my Docker install for a while (hours I guess?) resulted in a multi-gigabyte log file as well.

**Comment by tiborvass** [comment]
The UX of it is covered in #23706 . The fact that CPU goes up is specific to this issue. Moving to P2

**Comment by aluzzardi** [comment]
I believe this has been fixed with default restart delay.

Closing this down - @nathanleclaire are you okay with this?

**Comment by thaJeztah** [comment]
@aluzzardi which release is this fix in, is that in 1.12 or will be in 1.13? (so that I can set the correct milestone)

**Comment by nathanleclaire** [comment]
seems reasonable
